### PR TITLE
Stats_dist: don't use balance_last for multinomial distribution (#347)

### PR DESCRIPTION
**Pull request by pveber**
As described in #347, balance_last raises exceptions in case where
numerical inaccuracy prevent summation to 1 of input
probabilities. The comments there argue that not only the
transformation performed by balance_last, but is also not really
necessary. See notably this comment:

As I see it, owl's current implementation is very close to the one in
the GSL. It computes the sum of input probabilities (norm variable),
and it seems to me that it doesn't matter that these sum to 1. Playing
with gsl confirms it:

```ocaml
# Gsl.Randist.multinomial rng ~n:100000 ~p:[|1. ; 2.|];;
- : int array = [|33251; 66749|]
```

This seems very natural to me, and much less confusing and error prone
than ignoring the last element of the probability vector.

So in summary, it would be enough to remove the call to balance_last
in multinomial_rvs:

```ocaml
let multinomial_rvs n ~p =
  let k = Array.length p in
  let _p = Genarray.create float64 c_layout [|k|] in
  let _s = Genarray.create int32 c_layout [|k|] in
  Owl_utils.Array.balance_last 1. p |>
  Array.iteri (fun i a -> Genarray.set _p [|i|] a);
  _multinomial_rvs ~k ~n ~p:_p _s;
  Array.init k (fun i -> Genarray.get _s [|i|] |> Int32.to_int)
```
The same is true IMO for the pdf:
```ocaml
# Gsl.Randist.multinomial_pdf ~n:[|3;6|] ~p:[|1. ; 2.|];;
- : float = 0.273129096174363206
# Gsl.Randist.multinomial_pdf ~n:[|3;6|] ~p:[|0.333333333333333 ; 0.6666666666666|];;
- : float = 0.273129096174363706
```

### CODE REVIEW

File: src/owl/stats/owl_stats.mli [hunk 1]
```diff
@@ -492,11 +492,10 @@ from ``n`` trials. The probabilty mass function is as follows.
 .. math::
   P(x) = \frac{n!}{{x_1}! \cdot\cdot\cdot {x_k}!} p_{1}^{x_1} \cdot\cdot\cdot p_{k}^{x_k}
 
-``p`` is the probabilty mass of ``k`` categories. If the elements in ``p`` do
-not sum to 1, the last element of the ``p`` array is not used and is replaced
-with the remaining probability left over from the earlier elements.
+``p`` is the probabilty mass of ``k`` categories. The elements in ``p`` should
+all be positive (result is undefined in that case), but they don't need to sum
```
Comment (mseri) [inline src/owl/stats/owl_stats.mli]: What do you mean with `result is undefined in that case`?

**Comment by mseri** [comment]
Thanks for the PR!

**Comment by mseri** [comment]
Would it be worth to log a warning (containing the value of the sum) if the probabilities do not sum to 1? In this way we can still inform the user that something may be fishy

File: src/owl/stats/owl_stats.mli (hunk 1 above)
Comment (pveber) [reply src/owl/stats/owl_stats.mli]: Woops, sorry the formulation is very unclear. I meant: "result is undefined if not".

Well there's a tradeof here: either we check that all elements in `p` are positive and raise `Invalid_argument` if not; either we check nothing but don't give any guarantee on the result if the precondition is not verified (I have no idea what the function does in that case, although I'm confident it terminates). The thing is that checking `p` at each call can be very costly, and for large scale simulation code you wouldn't want to do that. Another possibility would be to introduce a private type of float array with only positive values, so that the check can be done once and for all.

**Comment by pveber** [comment]
I'm not too fond of logs, but anyway I think there is nothing fishy in providing an unnormalized set of (positive) weights: the underlying algorithm is just as good in that case, and it can be very convenient not to be forced to normalize the weights (and that's what the GSL does). In case you'd agree with that design, I should probably explain better in the doc :).

**Comment by pveber** [comment]
PS: I updated the initial message, it seems I had failed the intended copy/paste

**Comment by ryanrhymes** [comment]
Thanks for fixing this. The design looks very good to me. Logging can be supressed by increasing the logging level in owl so it should not be a problem. @mseri can we merge this?

File: src/owl/stats/owl_stats.mli (hunk 1 above)
Comment (pveber) [reply src/owl/stats/owl_stats.mli]: I added a commit to improve the description in the doc.

**Comment by mseri** [comment]
I like the latest docs... I was just commenting about the wording when you committed :)

**Comment by pveber** [comment]
Alright, thanks @mseri @ryanrhymes !
